File: gunhoflash/remove-nth-node-from-end-of-list.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function removeNthFromEnd(head: ListNode, n: number): ListNode | null {
  const headHead = new ListNode(0, head);
  let cursor = head;
  let delayedCursor = headHead;

  // Move cursor n - 1 steps forward
  while (--n) cursor = cursor.next!;

  // Move cursor and delayedCursor together
  while (cursor.next) {
    cursor = cursor.next;
    delayedCursor = delayedCursor.next!;
  }

  // delayedCursor is the node before the one to be removed
  delayedCursor.next = delayedCursor.next!.next;

  return headHead.next;
}

File: gunhoflash/remove-nth-node-from-end-of-list.py
from typing import Optional

class ListNode:
  def __init__(self, val = 0, next = None):
    self.val = val
    self.next = next

class Solution:
  def removeNthFromEnd(self, head: ListNode, n: int) -> Optional[ListNode]:
    headHead = ListNode(0, head)
    cursor = head
    delayedCursor = headHead

    # Move cursor n - 1 steps forward
    for _ in range(n - 1):
      cursor = cursor.next

    # Move cursor and delayedCursor together
    while cursor.next:
      cursor = cursor.next
      delayedCursor = delayedCursor.next

    # delayedCursor is the node before the one to be removed
    delayedCursor.next = delayedCursor.next.next

    return headHead.next
